File: content/snippets/scala2/interview_lcci/01/08_zero_matrix.scala
object Lcci0108ZeroMatrix {
  def setZeroes(matrix: Array[Array[Int]]): Unit = {
    val zeroRows = scala.collection.mutable.Set.empty[Int]
    val zeroCols = scala.collection.mutable.Set.empty[Int]

    for (r <- matrix.indices; c <- matrix(0).indices) {
      if (matrix(r)(c) == 0) {
        zeroRows += r
        zeroCols += c
      }
    }

    for (r <- matrix.indices; c <- matrix(0).indices) {
      if (zeroRows.contains(r) || zeroCols.contains(c)) {
        matrix(r)(c) = 0
      }
    }
  }

  def main(args: Array[String]): Unit = {
    val matrix = Array(
      Array(1, 2, 3),
      Array(4, 0, 6),
      Array(7, 8, 9)
    )

    setZeroes(matrix)
    matrix.foreach(row => println(row.mkString(" ")))
  }
}

File: content/snippets/typescript/interview_lcci/01/08_zero_matrix.ts
function setZeroes(matrix: number[][]): void {
  const zeroRows = new Set<number>();
  const zeroCols = new Set<number>();

  for (let r = 0; r < matrix.length; r++) {
    for (let c = 0; c < matrix[0].length; c++) {
      if (matrix[r][c] === 0) {
        zeroRows.add(r);
        zeroCols.add(c);
      }
    }
  }

  for (let r = 0; r < matrix.length; r++) {
    for (let c = 0; c < matrix[0].length; c++) {
      if (zeroRows.has(r) || zeroCols.has(c)) {
        matrix[r][c] = 0;
      }
    }
  }
}

function main(): void {
  const matrix = [
    [1, 2, 3],
    [4, 0, 6],
    [7, 8, 9],
  ];

  setZeroes(matrix);
  for (const row of matrix) {
    console.log(row.join(" "));
  }
}

main();
